src/notebooks/controllers/connectionTracker.ts

```diff
@@ -87,11 +87,15 @@ export class ConnectionTracker implements IExtensionSyncActivationService, IConn
      * & if this matches exactly with the last used kernel connection, then mark it as preferred.
      */
     @swallowExceptions()
-    private async onCreatedController(controller: IVSCodeNotebookController) {
+    private async onChangeController(e: IVSCodeNotebookControllerUpdateEvent) {
         await Promise.all(
-            workspace.notebookDocuments.map(async (notebook) => {
-                await this.updateAffinity(notebook, controller.connection);
-            })
+            e.added.map((controller) =>
+                Promise.all(
+                    workspace.notebookDocuments.map(async (notebook) => {
+                        await this.updateAffinity(notebook, controller.connection);
```

Comment:
Maybe we can have a batch affinity update impl.

Reply:
Will create a debt item for that.